Journal filtering

marusak opened the pull request:
Bunch of changes like text filtering, stopping/starting of streaming and possiblity to jump from page-journals to the the full journal.

![newlogsscreenreleasenotes](https://user-images.githubusercontent.com/12330670/80706391-8a909b80-8ae8-11ea-9a6f-4eca95255cfb.png)

![jumptyjumpy](https://user-images.githubusercontent.com/12330670/80706523-bca1fd80-8ae8-11ea-824b-f4e5a124a10a.png)

andreasn commented:
Great start!
For the follow on/off I'm thinking we might try another solution where it's on all the time, but we adapt the UI in such a way that it keeps the content view in place, and then message that there are newer messages available. A little bit how Slack and Discord does it.
https://support.discordapp.com/hc/user_images/S4X4hOZqvI_jkTGKIDw9Ww.png
I'll mock up a more detailed proposal for that.

andreasn commented:
Spotted some small little inconsistencies, such as that it's called prio in the URL, but priority in the filter. And that is the same as Severity, right? But smaller details like that, I think we can worry about later.

andreasn commented:
Also, I agree that we need to expand [the docs](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/managing_systems_using_the_rhel_8_web_console/reviewing-logs_system-management-using-the-rhel-8-web-console) to include the search syntax. Might be worth doing some inline, short docs for the basics. Like regex hints etc.

marusak commented:
> For the follow on/off I'm thinking we might try another solution where it's on all the time, but we adapt the UI in such a way that it keeps the content view in place, and then message that there are newer messages available. A little bit how Slack and Discord does it.

That is surely a good idea, but I think we want also stopping the following all together. We need to count with it streaming a loads of messages - which when connecting to machine through internet might not be wanted.

> Spotted some small little inconsistencies, such as that it's called prio in the URL, but priority in the filter. And that is the same as Severity, right?

Yeah, this is a bit annoying. We have kind of API, so I cannot just change it, I need to keep it in mind. And it is "prio". But I though that having it in more readable form in the input would be better. And yeah, it is the same as severity. Indeed not ideal. (I renamed "Service" to "Indentifier" to have the same name, but it is still "tag" in the url)
(https://cockpit-project.org/guide/latest/api-logs-html.html)

> Also, I agree that we need to expand the docs to include the search syntax

This is exactly what I had in mind. Once we finish this up, I'll get in touch with Lucka and hope we can improve on it.

>  Might be worth doing some inline, short docs for the basics. Like regex hints etc.

Sure, do you have some idea how it could look like and what it should say? (some rough idea is enough, we can polish it later).

larskarlitski commented:
This is amazing. I especially like that the date header stays on top. Great work!

I just talked to @andreasn and @garrett about the search syntax and have two cents to add :wink:. I it could be very useful to use the exact same filtering syntax that `journalctl` uses. One could copy the contents of the field into the terminal and the other way around­—for example to create the query in cockpit and then use it in a script.

Also, one would learn journalctl's syntax while using Cockpit :tada: 

I'm not sure if this is possible, but I wanted to throw my opinion into the discussion.

marusak commented:
> It could be very useful to use the exact same filtering syntax that journalctl uses. 

heh, that was my original intention as well, but when I got deeper into it, things were a bit more complicated, let me illustrate:

1. Firstly, all those arguments that I now use with ":" would need to start with "--", so instead of "priority:err" it would be "--priority=err"
2. "identifier:foo" is actually done through `SYSLOG_IDENTIFIER=foo` which would be bit confusing as it is not "--option"
3. "service:foo" is a bit crazy, it is interpreted as `_SYSTEMD_UNIT=foo + COREDUMP_UNIT=foo + UNIT=foo"` (yes, with '+' and spaces, which would make it awkward to parse as it is not clear if the "+" belongs to this filter, or you want to search for "+" (`_SYSTEMD_UNIT=foo +` means systemd_unit and look for "+" in message)
4. and my favourite is the time filtering. 'Previous boot' is '--boot=-1', 'Last 24 hours' is '--since=-1days', 'Last 7 days' is '--since=-7days', 'Recent' does literally nothing and 'Current boot' is `--boot`.
5. The text that you type to search is actually `--grep=<text>`

but apart from these thing it is almost 1 to 1 with journal ;)

garrett commented:
It's fine to have our own stripped down version of what you'd give `journalctl`; it would be nice if we could _also_ have a way to copy the exact `journalctl` commandline in a popup.

Simply having an input field would allow someone to copy and paste the exact string to see the journal entries to easily apply it to another machine. (Of course, they could _and_ should be able to bookmark and/or share the URL. But that's not easily copy/paste-able for multiple machines. We should probably have both.)

I'm thinking we have a (?) icon with a popover that stays open until you close it. And it would have a little bit of help there _and_ include a journalctl command line, ready for copy and pasting.

garrett commented:
BTW: This is _awesome_. The video really helps a lot too. Thanks for that! @andreasn and I were talking about it a lot in our call yesterday and (as you see above) we were both talking with @larskarlitski today, and we're all excited to get this in soon (in some form, with some changes).

marusak commented:
> Simply having an input field would allow someone to copy and paste the exact string to see the journal entries to easily apply it to another machine. (Of course, they could and should be able to bookmark and/or share the URL. But that's not easily copy/paste-able for multiple machines. We should probably have both.)

Both of these things are now working :) You can copy the whole search input and paste it into cockpit on different machine and it would filter with the same rules. Also the URL encodes it as well, so you can bookmark your current search or just change the host to see it on different machine.

> I'm thinking we have a (?) icon with a popover that stays open until you close it. And it would have a little bit of help there and include a journalctl command line, ready for copy and pasting.

Like that idea! Generating the journalctl command is not really difficult, it is harder just to parse back in if user can edit it.

garrett commented:
I mentioned to @andreasn in the call yesterday that we have two choices on how to implement filtering here:

1. As PF4 suggests
   - without a query bar
   - with lots of UI for toggling things
   - with tags that are dismissable
   - **Pros**:
      1. Consistency with other PF4 apps
   - **Cons**:
      1. Our users would have to click a bunch every time
2. Like GitHub
   - with a query bar
   - lots of UI for adding to the string
   - **Pros**:
      1. Copy/pastable between multiple machines
      2. Similar to `journalctl`
      3. We'd have presets of common, useful searches on the left of the input, like GitHub does
   - **Cons**:
      1. Inconsistent with PF4 apps

(There might be more pros and cons I might've forgotten. I tried to write this quickly to share the gist.)

The three of us are pretty heavily leaning toward the GitHub model.

andreasn commented:
Very, very rough sketch for how we could handle the flight of incoming messages:
![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/468e519578ce703a9767295a0a2d9a3dd88b6a9c/journal/incomingmessages.png)

andreasn commented:
Will attach designs for how to handle the syntax hints, journalctl command hints etc.

andreasn commented:
![mockup](https://raw.githubusercontent.com/cockpit-project/cockpit-design/741cd648ddd44b5443ea15a80cece54c15b48d8b/journal/inline-help.png)

andreasn commented:
Thinking we should move the inline help out to a separate Issue/PR.

andreasn commented:
I somehow managed to get myself into this state...

![Screenshot from 2020-04-21 14-19-58](https://user-images.githubusercontent.com/209895/79865341-495b0600-83db-11ea-96c9-699355ac2e0b.png)

marusak commented:
> I somehow managed to get myself into this state...

reproducer: 
1. Select 'Everything' in the severity dropdown
2. Delete everything in the search input and hit enter

andreasn commented:
One thing that become very evident with the filter box is how much the Identifier dropdown tend to jump so much (it does today too, but it doesn't push anything else around).
https://andreasn.portknox.net/s/XYerotgC5jTz54L

Not sure how to best fix this. Give it some min/max-width maybe? Or something with flex?

andreasn commented:
When I navigate from the logs of a service (for example accounts-daemon), it sets the string to `priority:debug service:accounts-daemon.service `, but the _Severity_ dropdown says _Error and above_

To reproduce:
1. Go to IP-ADDR:9090/system/services#/accounts-daemon.service
2. Click the _Go to all logs_ link

marusak commented:
@andreasn I fixed most of the issues we were talking about like the severity selector not changing, removed the following checkbox and put there button to pause etc.

Help and not showing new messages but having banner marking new messages would be followups. 

There is still this issue that it jumps a lot when services are loaded. Not quiet sure what we can do about it though.

Any other issues that I should work on? (based on the amount I may split this into more PRs, but lets firstly see, how much work is actually needs).

andreasn commented:
Thanks for the fixes!

I think I found another small inconsistent state. One that I think existed before this, but was not as evident. Happy to file that as a follow-up about it.

To reproduce:
1. Select current boot in the dropdown
2. Press the _Load earlier entities_ button (this will load things that are not the current boot)
3. Still says _current boot_, even though that is no longer true

marusak commented:
> I think I found another small inconsistent state. One that I think existed before this, but was not as evident. Happy to file that as a follow-up about it.

Yeah, this is also happening on the current master, so can you please open separate issue? Thanks!

andreasn commented on pkg/lib/cockpit-components-logs-panel.jsx:
```diff
@@ -130,7 +130,10 @@ export class LogsPanel extends React.Component {
-                <div className="panel-heading">{this.props.title}</div>
+                <div className="panel-heading">
+                    <h2 className="panel-title">{this.props.title}</h2>
+                    <button className="link-button" role="link" onClick={e => cockpit.jump(this.props.goto_url)}>{_("Go to all logs")}</button>
```
Thinking this doesn't need to be a verb. It's evident from the link styling that it will go to somewhere, so I would probably go for **All logs** instead.

marusak replied:
Done

andreasn commented:
Issues that are already in master, and independent of this PR:
Load more: #13954
Jumpy controls: #13955

andreasn commented on pkg/systemd/logs.html:
```diff
@@ -32,45 +32,46 @@
+      </span>
+      <div class="filters-toggle full-row" hidden></div>
+      <div class="filter">
+          <label for="journal-prio-menu" translate>Severity</label>
```
I wonder if this should be renamed to _Priority_, since that's the string we put in the text filter. A little weird if they are inconsistent with each other for no apparent reason.

marusak replied:
Done

marusak commented:
Pushed new version, rebased, cleanued up commits and fixed:
1. Rename 'Severity' to 'Priority'
2. Rename 'Go to all logs' to 'All logs'
3. Removed 'Everything' from 'Priority'
4. Fixed #13955

andreasn commented:
I think this is pretty good now!
Will play around some more with it, to see if I can uncover something, but I think it's good from a UX standpoint.

andreasn commented on pkg/systemd/logs.scss:
```diff
@@ -59,15 +59,67 @@
+
+.filter.text-search {
+    min-width: 150px;
+    max-width: 700px;
```
What is the reason for this line?
I think it's OK that the filter takes as much space as it wants.

Also makes the pause button not line up with the list below in some situations.
![Screenshot from 2020-04-23 14-29-32](https://user-images.githubusercontent.com/209895/80099743-409b3900-856f-11ea-8507-c36c565014ab.png)

marusak replied:
> What is the reason for this line?

Good point. Not much any more. I had it there when the screen was not wide enough to have things on one line, but wide enough that the search was on separate line it would span too much. But now it does nothing really. I'll remove it.

marusak replied:
Done

andreasn commented on test/verify/check-journal:
```diff
@@ -291,23 +307,99 @@ ExecStart=/bin/sh -c 'for s in $(seq 10); do echo SLOW; sleep 0.1; done; sleep 1
+        b.wait_present("#journal-box .cockpit-logline .cockpit-log-message:contains('Message 1')")
+
+        b.click("#journal-follow:contains('Pause')")
+        b.wait_visible("#journal-follow:contains('Stream')")
```
I demoed this to @garrett and @jgiardino today, and they suggested **Resume** might work better as a label than **Stream**, as it's a more common opposite verb to **Pause**, has clear analogues with playback in the physical world etc. Could even enforce that with the icons fa-play and fa-pause.
This is also what OpenShift does for their log streaming UI, so it's a gain for cross-product consistency.
I'll try to find a screenshot of that.

andreasn commented on pkg/systemd/logs.js:
```diff
@@ -915,25 +913,94 @@ $(function() {
+    $('#journal-follow').on("click", function() {
+        const state = $(this).attr("data-following") === "true";
+        if (state) {
+            $(this).text(_("Stream"));
```
I demoed this to @garrett and @jgiardino today, and they suggested Resume might work better as a label than Stream, as it's a more common opposite verb to Pause, has clear analogues with playback in the physical world etc. Could even enforce that with the icons fa-play and fa-pause.
This is also what OpenShift does for their log streaming UI, so it's a gain for cross-product consistency.
I'll try to find a screenshot of that.

marusak replied:
Makes sense, updated!

marusak commented:
Hmm, [firefox](https://logs.cockpit-project.org/logs/pull-13905-20200426-142919-1cab9aab-fedora-31-firefox/log.html#211-2) has hard time to set up priority. This seems rather strange, needs to be looked at.

KKoukiou commented on test/verify/check-journal:
```diff
@@ -337,6 +337,7 @@ ExecStart=/bin/sh -c 'for s in $(seq 10); do echo SLOW; sleep 0.1; done; sleep 1
             b.click("#journal-service-menu button")
             b.wait_not_present("#journal-services-list ul li a:contains('just-a-service')")
         else:
+            wait_log_not_present("WORK PLEASE")
```
:smile:

KKoukiou commented on pkg/lib/cockpit-components-logs-panel.jsx:
```diff
@@ -130,7 +130,10 @@ export class LogsPanel extends React.Component {
-                <div className="panel-heading">{this.props.title}</div>
+                <div className="panel-heading">
+                    <h2 className="panel-title">{this.props.title}</h2>
+                    <button className="link-button" role="link" onClick={e => cockpit.jump(this.props.goto_url)}>{_("All logs")}</button>
```
If this property is not passed maybe don't show the button?
Else make sure that all places that use that Component pass that prop, and mark it as required in PropTypes.

marusak replied:
fixed

KKoukiou commented on pkg/systemd/logs.html:
```diff
@@ -47,7 +47,7 @@
       <div class="filters-toggle full-row" hidden></div>
       <div class="filter">
-          <label for="journal-prio-menu" translate>Severity</label>
+          <label for="journal-prio-menu" translate>Priority</label>
```
+1 for following the wording from the manual as well.

garrett commented on pkg/systemd/logs.html:
```diff
@@ -45,24 +45,18 @@
-        </ul>
-      </div>
+      <label for="journal-prio-menu" translate>Severity</label>
+      <select id="journal-prio-menu" class="ct-select">
```
I 100% agree with @KKoukiou here.

Just a little extra info: Our original ct-select was a mess, but we redid it to be based on a native form element. It was to fix old code and as a stopgap measure before moving to PF4 selects. I agree that we should prefer PF4's form-based control when possible.

For HTML, it's exactly as @KKoukiou says, a "form control", specifically the select: https://www.patternfly.org/v4/documentation/core/components/formcontrol#select

In React, there are a lot of "selects", but the component we generally want is "form select": https://www.patternfly.org/v4/documentation/react/components/formselect#component-title

(PF4 doesn't map these two 1:1 between HTML and React and there are so many other things they call "selects" that are different that I thought it made sense to point out all of this above.)

garrett replied:
I tried `pf-c-form-control` and had unstyled forms. I then realized that PF4 isn't being brought in. Until we pull in PF4 into the logs, I guess `ct-select` is OK.

KKoukiou replied:
Pulling PF4 FormSelect css into the logs is one line import, not such a big deal.

garrett commented on pkg/systemd/logs.scss:
```diff
@@ -80,13 +80,32 @@
+}
+
+.filter.text-search {
+    min-width: 150px;
```
Yeah, we shouldn't use px for sizing and spacing when we can avoid it. It's necessary in media queries (as some browsers don't support rem/em), useful for 1px borders, and we sometimes base icon sizes on it, but otherwise, it's best to not use it.

In this case, it would be better to use em instead of rem, as rem is the root em (the base font size) whereas em is based on the the local font size. If someone is trying to say "it's about this long with the text string", you generally want something like em (or ex or ch), which is based on the character size.

marusak commented on pkg/systemd/logs.js:
```diff
@@ -863,6 +864,18 @@ $(function() {
 
+    function fit_filters() {
+        const filters = $("#journal .content-header-extra .filter");
+        if ($(filters[filters.length - 1]).offset().top > 20) {
```
Well, the thing is that css screen sizes won't work here. We need to be absolutely sure when things started to wrap (because we want to keep the date sticky to the top and having two rows would break that).
We can set 'max-width' to every single element, but that will break things when translations are used. I think we are better off to just check if the last element broke to another line and then just apply the logic, than to try to guess screen sizes and when it might break in which languages...
The hardcoded `20` just means that the element is further than 20px from the top of the parent - which happens only when it wraps. (margin is 10px)

garrett replied:
I guess I should look at this. It really has bad "[code smell](https://en.wikipedia.org/wiki/Code_smell)", as it's doing presentation-related logic in JS instead of using CSS.

You shouldn't have to:
- use a string of selectors in the JS
- care about pixel values (which are easily messed up with browser zoom / text zoom — and browser zoom is very common; recent Firefox releases even support global zooming across all sites).

garrett replied:
Note: Even changing the OS to large fonts — a setting completely independent of the browser — will impact things in ways you won't expect.

(You may need to restart a browser after enabling "large text" in GNOME's settings, or something similar on other desktops. Firefox doesn't change the fonts on the fly, but Chrome seems to.)

marusak replied:
This has been changed all together.

garrett commented:
Is the PR supposed to add all filtering behind a toggle link?

![Screenshot_2020-04-28 Logs - Rain](https://user-images.githubusercontent.com/10246/80465843-01d4fc80-893c-11ea-9824-ab7bce475bfd.png)

marusak commented:
> Is the PR supposed to add all filtering behind a toggle link?

Only when it does not fit on the first line. If the screen is wide enough, it is not hidden. (it works when you resize down, but resizing up does not jump back to the "just one line", you need to reload. But I guess that it is fine, as users usually don't resize up and down, and even if someone would decide to make it bigger, not that big of a deal, that it would not shown up).

> I'm not sure what is supposed to happen and what we'd ideally want to happen, especially as this is different from what I expected based on the screenshots and quickly reading the conversation above. Perhaps something was discussed on IRC that wasn't captured here in the PR?

I explained it in the video (that is now obsolete, but this idea did not change). The thing is - we want to collapse it behind toggle link when it would not fit on one line. The main reason is that we have the date sticky, so we can implement it easily. The other reason is that on phone it can take 5 lines, which is significant amount of the screen taken unnecessary.

garrett commented:
I still think we have a few ways to implement the UI — and what I see in this PR doesn't follow any of the conventions. I'll detail the two solutions we talked about in IRC, in calls, and such, but haven't been represented in mockups or in the PR yet.

---

### Solution 1: A row of filters

The PF page on filter design (https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/filters) has several different options.

The simplest one is what they call "[multiple filters](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/filters#multiple-filters)" — but this doesn't have a text entry component. It looks like this:

![3e4196ada6167c2745cf79d3c1feb448-multiple-filters](https://user-images.githubusercontent.com/10246/80469449-dbfe2680-8940-11ea-80c3-c6b07a44c11f.png)


If we do want to toggle visibility of some advanced widgets, I'd suggest we always show as many filter widgets as possible, using the pattern linked above, but toggle the query string and the popover to get additional help and the systemd command line.

Toggling should set a cookie so the toggle state will always be the state last selected in that particular browser. (And, of course, it should default to the simple mode without the text entry field.)

---

### Solution 2: Expression builder

Alternatively, there's the "expression builder", which isn't shown or described whatsoever on PF4's page. However, I'm guessing it looks like GitHub's issue search. Here's a screenshot:

![Screenshot_2020-04-28 cockpit-project cockpit](https://user-images.githubusercontent.com/10246/80469238-92154080-8940-11ea-86de-c6aeca1a7061.png)

We'd probably want to collapse that to one line (when it fits), however.

[ Presets ▾ ][_______________]\(i)   [ Time ▾ ] [ Priority ▾ ] [ Identifier ▾ ]  |  (Pause)

This would be either:
- grid (with breakpoints): auto 1fr repeat(6, auto)
- flex, with the text entry set to `flex: auto` (or perhaps 1 0 auto, so it'll grow, but not shrink), with the whole thing set to flex-wrap, and probably grouped with presets+entry+1 and the reset as another group 

(Flex makes more sense here.)

Presets would include the most useful searches, with possible things like:
- Current boot
- Previous boot
- Crashes
- sudo (priority: debug, identifier: sudo)
- SELinux (priority: debug, identifiers: setroubleshoot, setroubleshootd)
- etc.

garrett commented:
> The thing is - we want to collapse it behind toggle link when it would not fit on one line.

Sure, but the widgets chosen to hide aren't the right ones.

> The main reason is that we have the date sticky, so we can implement it easily. 

Right, but the date doesn't show up when the area is collapsed.

> The other reason is that on phone it can take 5 lines, which is significant amount of the screen taken unnecessary.

Right, but this is also a problem on a very reasonably large desktop machine, especially when multitasking (as system admins often do).

andreasn commented:
> Presets would include the most useful searches, with possible things like:
> 
>     Current boot
>     Previous boot
>     Crashes
>     sudo (priority: debug, identifier: sudo)
>     SELinux (priority: debug, identifiers: setroubleshoot, setroubleshootd)
>     etc.
> 

I think these are great suggestions, but it tacks more stuff onto an PR already full of various stuff. ~500 lines of changes already. In the situation we mess something up and have to back something out of master again, we'll have to back all of these changes out at once. If we break out this suggestion into a separate PR instead, we'll have plenty of time to discuss exactly how it should behave, without blocking the fixes in this PR from landing. I had really wished we could land this PR before the next release, because I think it's good if we can get feedback on it early and allow it to mature gradually over time.

garrett commented:
> I had really wished we could land this PR before the next release, because I think it's good if we can get feedback on it early and allow it to mature gradually over time.

Sure, so it would make sense to scale back this PR and decide on the direction we'd like to go. I'm not talking about all of this in this PR right now, I'm talking about the design on where it should go. If we do this ad-hoc here and there, we'll never get to a good design.

The first option is not entirely unlike a hybrid of what we currently have and this PR. It probably wouldn't take much effort to change it to be more like that. We could decide — later — that we want to have option 2, where the filter bar is more prominent.

Right now, this PR hides all the filtering on most monitor sizes, except for excessively large screens (which many of _us_ might have) with a fullscreened browser. That's a UI regression. This PR isn't ready to land right now, unless that's solved first.

My suggestion was to show the filtering widgets and hide the filter bar by default, instead of hiding everything by default. And it's also to move the labels above the form elements, as PF4 suggests too, which provides more horizontal space, which means less of the need to hide elements.

Again, we could, later, decide on a query interface, but we should know what target we should design toward.

garrett commented:
(Also, what we talked about in IRC and in a few standups was _not_ documented anywhere.

As it was relevant to this PR, I documented it here, so we could refer to it in the future, either with a copy/paste, or linking. I didn't intend for it to be completed with this PR, but we should actually talk about it and record things when we do. Hence, me documenting it here.)

marusak commented:
@garrett since majority of your required changes are css stuff (like hiding "correct" elements, showing more elements when things are collapsed, having labels above inputs...) and also not using `px` but `em`, would you be interested to send patch so it does not cause "UI regression"?

As @andreasn mentioned, I really don't want to get any more changes here, it is already 20 commits and reworking the whole filtering and searching logic is just for a different day.

garrett commented:
Anyway, to succinctly summarize, the MVP (minimum viable pull-request) would need the following:

- [ ] do not hide the filtering widgets by default
- [ ] show the date at the top, even when expanded (if show/hide stays in)

We could probably achieve this by reworking the layout with the labels on top like the PF4 mockup.


Time | Priority | Identifer | Text |   |   |  
:-- | :-- | :-- | :-- | :-- | -- | --
[ Recent ▾ ] | [ Warning and above ▾ ] | [ All ▾ ] | [_____] | Reset filters | \| | Pause

garrett commented:
(Note: I've jumped in and am looking at these layout and CSS related things.)

garrett commented:
I've addressed the issues.

For small sizes (incl. mobile, but also very narrow desktop browser windows), I am using the show/hide link to show all the filtering or none. It impacts the size of the viewable area, but it's easy enough to toggle. (The list is then considered a preview, basically, I suppose...)

By removing the link on larger sizes, we have more room for the filtering widgets. For the filtering widgets on desktop, they re-flow if needed, based on content sizes. The labels also stack on top in most cases, or slide over to the left in mobile/small (to preserve vertical space) and on huge layouts (when there's more than enough room), providing more space for the list.

garrett commented:
@marusak and I decided that this is feature-complete for the PR and all the mockups and designs and feature enhancements spoken about that aren't in code are suitable for consideration in follow-up PRs. (We might or might not do them.)

Meanwhile, what we have (aside from a small bug or two) is basically ready for merging (once those small bugs are fixed and tests are updated).

garrett commented:
It's weird without the label:

![Screenshot_2020-04-30 Logs - Rain](https://user-images.githubusercontent.com/10246/80703276-c7f22a80-8ae2-11ea-9637-be86ba7f971b.png)

marusak commented:
> It's weird without the label:

Yeah, I did the rebase wrong, need to fix it.

[This](https://github.com/cockpit-project/cockpit/compare/1e07eca31d8f18b27efcd621caa5550df61d0463..a20752b4eb14a2703bbab76848367eb0454601d2) should be empty.

garrett commented:
I went back to my branch before pulling and ran cherry on it:

```
$ git cherry master -v
+ 03b8c385d87f2966305e9af6cb7acdb51bd3096d logs: Use '<select>' for priority
+ 3feed597ffe42df485495782d4ab4ee2f8ba52b8 logs: Filter priority with `--priority` and not `PRIORITY` matching
+ 1106ca60679fb929f07db2d75b619986c854042e logs: Indicate current day by keeping panel heading sticky
+ 830cff63f71433df1b347ea80d52265f6b2f44ed logs: Use '<select>' for time and set current value
+ 3b326b44dfcf14a7d493c04fc33e7f9e3acc7cac doc: Document also 'start=previous-boot' system API
+ ec727bc2fe1e8a3f68d927e673c92b06c3397fba test: Logs do not support 'start=oldest'
+ 2e7162f4a852ddb985a03aa533afcfb4345e4909 logs: Use '<select>' for filtering by service
+ 7884203469836c4719d7d5cbf0f3cec9481494f8 logs: Improve mobile mode
+ bf37b76f0b0c24e5ec8e9e5f53c024ca7c7aa469 logs: Enable to stop following (and start it back up)
+ a3e8aaaef8c9081063c26bb449a7650187efbe0b logs: Enable to filter by message
+ 61138d318997c2a2e20982aeada9b04f83630845 logs: Enable passing miscellaneous filters
+ cb87df7e609be1c98cb95c24c2e60acfd0c6be9c storage,network: Enable going from logs panel to full logs
+ fcfe8874888c65e7a15fc91cfec4915843f72f12 logs: 'service' option shows all related info about the service
+ 628fff90e9e314ede6be99695370fad7d58cef65 services: Enable going from service logs to logs page
+ e4caaf70896bcac4bcd5ad4151e15a3426f1307a logs: Rename 'Service' label to 'Identifier'
+ bf9d00d0e1e9dfe0af9ca00286084d3f031fa5cc logs: Show all filters in the search box
+ fd40dc659e587c5d2b3b8d46e50d2cf433bed94f logs: Fix width of service selector
+ b17dbce0e668807556234b505430dc02208c85b6 services: Enable going from service logs to logs page
+ 73d9befe6c88c4180ff5e2981c4398920d4aee29 logs: Rename 'Severity' label to 'Priority'
+ 1cab9aab0f4a66a2ebc4170a74807ce2d04ab0e3 logs: Drop 'Everything' option from priority
+ 3a9b911ec1f5e506681e28b8734f025e5388b5a0 logs: Rework filter layout
+ 28ed39978ba8802cdbb1bbfe848929fcb8a0d0a2 logs: Make list scroll; fix sticky date headings
+ 8dde0f0ac0b5f70e20ec50dbcc3c0c2466b8f550 logs: Restyle headings
+ f987937153ccf023835d3f811eef3c250c33546d logs: Adjusted show/hide behavior & tweaked layout
+ 0ff0b80b5eebe17aa017ccedc0d50db7a0aa982c journal: Never wrap the time
+ 4355ee4ea30a32fa812996002b5f37e18f630e30 logs: Add extra padding to the sides of log lines
+ 1e07eca31d8f18b27efcd621caa5550df61d0463 logs: Add spacing above headers, remove from below
```

garrett commented:
Also,

```
git log --oneline | head
1e07eca31 logs: Add spacing above headers, remove from below
4355ee4ea logs: Add extra padding to the sides of log lines
0ff0b80b5 journal: Never wrap the time
f98793715 logs: Adjusted show/hide behavior & tweaked layout
8dde0f0ac logs: Restyle headings
28ed39978 logs: Make list scroll; fix sticky date headings
3a9b911ec logs: Rework filter layout
1cab9aab0 logs: Drop 'Everything' option from priority
73d9befe6 logs: Rename 'Severity' label to 'Priority'
b17dbce0e services: Enable going from service logs to logs page
```

marusak commented on test/verify/check-services:
```diff
@@ -261,6 +261,14 @@ Unit=test.service
         b.wait_in_text('.cockpit-log-panel .panel-body', "START")
         b.wait_in_text('.cockpit-log-panel .panel-body', "WORKING")
 
+        if m.image != "rhel-8-2-distropkg": # Introduced in #13905
```
Done

garrett commented:
Oddly, I have this difference between our branches now:

```diff
diff --git pkg/lib/cockpit-components-logs-panel.jsx pkg/lib/cockpit-components-logs-panel.jsx
index e1ff58ed5..e156aa29b 100644
--- pkg/lib/cockpit-components-logs-panel.jsx
+++ pkg/lib/cockpit-components-logs-panel.jsx
@@ -132,7 +132,7 @@ export class LogsPanel extends React.Component {
             <div className="panel panel-default cockpit-log-panel" role="table">
                 <div className="panel-heading">
                     <h2 className="panel-title">{this.props.title}</h2>
-                    { this.props.goto_url && <button className="link-button" role="link" onClick={e => cockpit.jump(this.props.goto_url)}>{_("All logs")}</button> }
+                    <button className="link-button" role="link" onClick={e => cockpit.jump(this.props.goto_url)}>{_("All logs")}</button>
                 </div>
                 <div className={"panel-body " + ((!this.state.logs.length && this.props.emptyMessage.length) ? "empty-message" : "")} role="rowgroup">
                     { this.state.logs.length ? this.state.logs : this.props.emptyMessage }
diff --git pkg/systemd/logs.scss pkg/systemd/logs.scss
index f4b109cdc..d871dcaab 100644
--- pkg/systemd/logs.scss
+++ pkg/systemd/logs.scss
@@ -191,10 +191,6 @@
     width: 8em; // Static width as it likes to resize a lot while loading
 }
 
-.filters-toggle {
-    flex-grow: 1;
-}
-
 .filters-toggle button {
     white-space: nowrap;
 }
```

garrett commented:
> oh right, the one in cockpit-components-logs-panel.jsx is on purpose

OK!

>  the one in logs.scss slipped through, but it actually does not have any affect. I'll drop it if this needs another push.

Yeah, I think I purposefully removed it at some point in some commit.

KKoukiou commented:
Retriggering [debian-stable](https://logs.cockpit-project.org/logs/pull-13905-20200430-145209-e3c636ac-debian-stable/log.html#253-2) because of one failing test.

KKoukiou commented:
@marusak seems https://logs.cockpit-project.org/logs/pull-13905-20200505-075636-e3c636ac-debian-stable/log.html#262-2 failed for a second time in a row.

marusak commented:
right, `debian-stable` does not support grep:
```
journalctl --grep 1
Compiled without pattern matching support
```

There is issue about it: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1751006

So I guess just override in tests will be enough?

garrett commented:
> right, debian-stable does not support grep

Oof.

Can we detect the cases where grep isn't supported and hide the text box? (I know it has the filtering there too, but it's misleading to have an input box without arbitrary string support.)

Meanwhile, the distributions where it's not supported will shrink over time, so it becomes less as an issue as time marches on.

marusak commented:
> Can we detect the cases where grep isn't supported and hide the text box?

We might want to just disable it? We still need it when users navigate from journal boxes to show, that there is actually some filter. Otherwise it is plainly confusing.

marusak commented:
different idea: What about showing alert box when user tries to use this grep functionality and it fails?
Detection is not that easy, I would need to do some dummy query first and then hide/disable it. With just warning user can have the box for other stuff and be notified when actually tries to use unsupported function. It is also much easier to implement.

garrett commented:
How many distros is this an issue for? Debian stable, Ubuntu 19.04 and before... and...??? How many of these are new releases of Cockpit coming out for?

Can we just special case it in the code, as we cannot properly detect it?

Showing something and then forbidding people from using it is really quite bad UX. An alert box is even worse. It's not an error that they caused, nor is it something that needs confirmation.

marusak commented:
> How many distros is this an issue for? Debian stable, Ubuntu 19.04 and before... and...??? How many of these are new releases of Cockpit coming out for?

From what we support just debian-stable.

> Can we just special case it in the code, as we cannot properly detect it?

uff, not easily. It would need to be some special case while we package stuff. Not really sure if we do something like that for js.

> Showing something and then forbidding people from using it is really quite bad UX. An alert box is even worse. It's not an error that they caused, nor is it something that needs confirmation.

Well, it would be like - `unsupported version` warning. The same as journalctl does it, it supports the `--grep` option, documents it and shows it in `--help`, just when you use it, it will say that your version is compiled without it. We would still need some time until we would recognize if it is usable - and thus we would firstly show the box and then hide it or we would not show it by default and then it would appear.

And I still feel like that hiding it is not a great option as when user comes from services page through `All logs` link, there will be no indication that some filters are applied.

garrett commented:
>>    Can we just special case it in the code, as we cannot properly detect it?

> uff, not easily. It would need to be some special case while we package stuff. Not really sure if we do something like that for js.

No, I mean, since it's a runtime issue, not a packaging one — and since it's only for Debian <=10.3, I mean adding a check.

I suppose we could do something like the following then:

Pseudocode:

```js
If (distro == debian && distro_version < 11) {
display_warning_banner_below_the_filter_bar();
}
```

Then we can have a notice at the top that says:

> Journaling in Debian 10 does not have search string support. [More information](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=890265)

marusak commented:
> No, I mean, since it's a runtime issue, not a packaging one — and since it's only for Debian <=10.3, I mean adding a check.

Unfortunate not. It is not related to the OS, just to the specific build. You can installe newer version from backports and have it working on debian < 11.
Also we cannot neither check version of systemd, as this version has this option, just the specific build does not include it.
Also we cannot check if `--grep` option is available, as it is is available, but not supported in the given build.

martinpitt commented:
Could we just check for `+PCRE2` in `journalctl --version`? At least it's on on Fedora, and off (`-PCRE2`) on my Debian 10 (stable) server.

marusak commented:
Fixed. Tests are reasonably green without single retry.
